### PR TITLE
RockBotBoss Designed and Warden fixes

### PR DESCRIPTION
**Pull request by koyomi7**
## Rock Robot Boss

- [x] **Summary**
This boss serves as the final enemy the player is going to face at a certain level. It should only be this one entity in a boss room
I manually changed the pivot points for multiple warden sprites to make it look better, it does not teleport now, it stays there and serves better as a roadblock.
- [x] **Methods Implemented**
**In the boss state machine**
- EventTrigger() modified so it has special attack effects on the player.
- ReducedMovement() in PlayerController.cs modified so it contains the method I used for one of the boss's attack to slow the player.

- [x] **Fields Implemented**
- In PlayerController.cs, added currentSpeed variable 

- [x] **Changes to Unity Editor (Screenshots, if applicable)**
- I disabled Boss in the editor since it takes lots of space, I left warden standing there like a chad cause I like him.

- [x] **Related Issue**
Nah, I'd win

- [x] **Mention Reviewers**
@YordyCruz @voxelit @Haikp 

- [x] **Miscellaneous & Images**
- Warden Sprite Fixes
wardenupdate-ezgif com-video-to-gif-converter
- **Boss Details**
- Beam attack, light knockback and huge damage 
boss-ezgif com-video-to-gif-converter
- Shoot Attack, light knockback and medium damage
boss-ezgif com-video-to-gif-converter (2)
- Charge attack, heavy knockback and light damage
boss-ezgif com-video-to-gif-converter (1)
- Slam attack, light knockback, light damage, slows player for a period of time.
boss-ezgif com-video-to-gif-converter (3)

### CODE REVIEW

**Comment by YordyCruz** [comment]
I love the Rock Robot Boss design this makes the boss encounter more engaging for the player gaming experience.

-  I have a few suggestions and questions about the boss:

     1. When testing the player can walk through the boss, and the boss can move into the player when attacked. Is this intended or could we implement collision handling to prevent this from happening?
     
     3.  When the player's projectile hits the boss, it sometimes registers multiple hits. Could we look into adjusting the hit detection to ensure it's registering as intended?
     
     5. If the player can spam attacks, then the boss might not transition to an attack animation. This might cause an issue especially if it’s just the boss in the room. Could we adjust the bosses to handle this scenario, maybe having a invulnerability Phases?
     
     7.  It would be great if the RockRobotBoss prefab could be added to the Prefab folder for consistency and easy access. Could you take care of this?

Thank you for your hard work on the Rock Robot boss and the Warden sprite fixes . These additions are making the game even more thrilling to play.

**Comment by Haikp** [comment]
A slight problem that could happen is two of the attacking being partially unfair. Charge attack and slam attack seem to have a much larger hitbox than what is supposed to be there. Atleast in the case of the slam attack, there is an animation, but the charge attack doesn't have much counter play, and the player will get hit. Perhaps we could change it to something else? We have discussed it to be a self buff. It would be pretty funny for the boss to buff himself 5 times in a row and just one shot the player.

File: 2DRove/Assets/Scripts/StateMachines/RockRobotBossFSM/RockBossAttackState.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                    collider.GetComponent<Rigidbody2D>().AddForce(-knockbackDirection * 5, ForceMode2D.Impulse);
+                    collider.GetComponent<Animator>().SetTrigger("Hit");
+                    playerScript.ReduceMoveSpeed(35, 1.5f);
+                }
+                
+            }
+        }
+        
+    }
+    public override void TakeDamage(RockBossStateManager RockBoss)
```
Comment (Haikp) [inline 2DRove/Assets/Scripts/StateMachines/RockRobotBossFSM/RockBossAttackState.cs:110]: A big problem here, and this goes for all the states, is that I believe bosses should not have hitstun. They could just be permanently hit over and over again and just have their attack get cancelled. I really do thing its best to just not have any hitstun. Simply moving over to the Hit State would interrupt his attack, so just dont have him go over to the hitstate, and instead when calling TakeDamage, just check if the boss' HP < 0.

File: 2DRove/Assets/Scripts/StateMachines/RockRobotBossFSM/RockBossAttackState.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    public override void UpdateState(RockBossStateManager RockBoss)
+    {
+        if(attackTime <= 0)
+        {
+            animator.SetBool("Beaming", false);
+            animator.SetBool("Shooting", false);
+            animator.SetBool("Charging", false);
+            animator.SetBool("Spinning", false);
```
Comment (Haikp) [inline 2DRove/Assets/Scripts/StateMachines/RockRobotBossFSM/RockBossAttackState.cs:50]: Remove this line, you took this bool out already.

**Review by Haikp — CHANGES_REQUESTED** [review]
There are some changes that I have recommended, the hit state one in particular should be changed for all the rockboss' states.

Also, remove that one SetBool in attackState.

Aside from that, good job, some fine tuning can be done once we get to try to fight the boss  in a better setting.

File: 2DRove/Assets/Scripts/StateMachines/RockRobotBossFSM/RockBossAttackState.cs (hunk 2 above)
Comment (koyomi7) [reply 2DRove/Assets/Scripts/StateMachines/RockRobotBossFSM/RockBossAttackState.cs:50]: noted, will change very soon

**Comment by voxelit** [comment]
The Rock Bot Boss looks very nice! Good work Mr. Henry.

I really like how different it is from the regular mobs we currently have, with the different types of attacks and adding status effects to the player. Good job on implementing these features.

I did come across an issue where the boss would teleport when hit by the player's attack. This could potentially lead to the game becoming unplayable because the boss would teleport out of bounds. It would be appreciated if you could fix this.

Screenshot 2024-03-31 193300

Otherwise, keep up the great work!

**Pull request merged**